---
title: "Vignette1_HuMMuS"
author: "Trimbour RÃ©mi"
date: "2023-05-16"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE }
knitr::opts_chunk$set(eval = FALSE)
devtools::load_all("../hummus_package")
```
## Import single-cell data

```{r}
scRNA <- read.table("data/real_example/hESC_Chen_scRNA.tsv")
scATAC <- read.table("data/real_example/hESC_Chen_scATAC_bin.tsv")
scRNA <- scRNA[1:700, ]
scATAC <- scATAC[1:1000, ]
```

## Transform in hummus object
```{r}
# Create seurat object with scRNA-seq data
seurat <- SeuratObject::CreateSeuratObject(scRNA)
# Add scATAC-seq data to seurat object
seurat[["peaks"]] <- Signac::CreateChromatinAssay(scATAC, sep = c(":", "-"))

# Create an hummus object from seurat object
hummus <- hummus_object(seurat)
```

# Add genome annotations to hummus object
Fetch genome annotations
```{r warning=FALSE}
genome_annotations <- get_genome_annotations(
  ensdb_annotations = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86)
```
Add genome annotations to seurat object
```{r}
Signac::Annotation(hummus@assays$peaks) <- genome_annotations
# Load TF motifs from JASPAR2020 and chromVARmotifs in hummus object
hummus@motifs_db <- get_tf2motifs()
```

# Compute bipartites and add it to hummus object
Add bipartite between rna-genes and peaks data to hummus object
```{r}
hummus <- bipartite_peaks2genes(
                      hummus_object = hummus,
                      gene_assay = "RNA",
                      peak_assay = "peaks",
                      store_network = FALSE,
                      )
```

Add bipartite between tfs and peaks data to hummus object
```{r}
hummus <- bipartite_tfs2peaks(
              hummus_object = hummus,
              tf_expr_assay = "RNA", #use to filter TF on only expressed TFs, if NULL, all TFs with motifs are used
              peak_assay = "peaks",
              tf_network_name = "TF_network",
              genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
              )
```

# Compute layer networks and add it to hummus object

Compute the TF network from OmniPath database
```{r}
hummus <- compute_tf_network(hummus,
                            gene_assay = "RNA",
                            verbose = 1,
                            multiplex_name = "TF",
                            )
```


Compute and save gene network from scRNA-seq w/ GENIE3
```{r}
hummus <- compute_gene_network(hummus,
                              gene_assay = "RNA",
                              method = "GENIE3",
                              verbose = 1,
                              number_cores = 5,
                              store_network = TRUE,
                              output_file = "gene_network.tsv")
```

Compute the peak network from scATAC-seq w/ Cicero
```{r}
#hummus <- compute_atac_peak_network(hummus)
```

# Save the mulilayer in a classic hierarchical structure
```{r}
save_multilayer(hummus = hummus, "a")
```

