---
title: "Predict GRN from scRNA+scATAC data (Chen 2018 dataset)"
author: "Trimbour Rémi"
date: "2023-05-16"
output:
  github_document: default
  html_document: default
  pdf_document: default
---


```{r setup, include=FALSE}
#knitr::opts_chunk$set(eval = FALSE)
devtools::install_github("cantinilab/HuMMuS")
```
## General description of the vignette
##### 0. Preparation of the environment
##### 1. Initialisation of a hummus object 

##### 2. Construction of the multilayer
  We want to link TFs to peaks (atac regions), then peaks (atac regions) to 
  genes. We thus need 2 bipartites:
  
  - 2.1. **TF - peaks bipartite**
  - 2.2. **peaks - genes bipartite**
  
  We also need one layer per type of features : TF, peaks and genes.
  
  - 2.3. **TFs/Proteins layer**
  - 2.4 **Peaks/ATAC layer**
  - 2.5 **Genes/RNA layer**
  
##### 3. Analyse multilayer and define gene regulatory network (GRN)$
  This part requires first have hummuspy python package (pip install hummuspy)
  installed, and second to save locally the multilayer files
  
  - 3.1. Save the network files
  - 3.2. Create GRN output
  - 3.3. Create enhancers output
  - 3.4. Create binding regions output
  - 3.5. Create target genes output
    

use_condaenv("base")

## Download the single-cell data
The data used in this tutorial can be downloaded at : ibens repo link

## 0. Setting up the environment
```{r}
library(HuMMuS)
library(reticulate)
```

## 1. Initialisation of HuMMuS object
HuMMuS R objects are instances developed on top of seurat objects. It means it’s created from a seurat object and the contained assays can be accessed the same way.

Additionally, it contains a motifs_db object, providing tf motifs informations, and a multilayer objects, that will be completed while going through this tutorial. It will mostly include :

  - list of multiplex networks (one per modality) 
  - list of bipartites (one per connection between layers)
  
```{r}
scRNA <- read.table("../data/real_example/hESC_Chen_scRNA.tsv")
scATAC <- read.table("../data/real_example/hESC_Chen_scATAC_bin.tsv")
# subsetting data to infer faster a toy case : specific for the tutorial, dont keep it in real cases !!
scRNA <- scRNA[1:700, ]
scATAC <- scATAC[1:1000, ]
```

#### 1.1. Transform data into a hummus object
```{r}
# Create seurat object with scRNA-seq data
seurat <- SeuratObject::CreateSeuratObject(scRNA)

# Add scATAC-seq data to seurat object
seurat[["peaks"]] <- Signac::CreateChromatinAssay(scATAC, sep = c(":", "-"))

# Create an hummus object from seurat object
hummus <- as(seurat, 'hummus_object')
```

#### 1.2. Add genome annotations to hummus object
Fetch genome annotations
```{r warning=FALSE}
# get human genome annotation from EndDb data
genome_annotations <- get_genome_annotations(
  ensdb_annotations = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86)
```
Add genome annotations to hummus/seurat object
```{r}
# We add annotation to the ATAC/peak assay
# (will be used to define peak-gene links)
Signac::Annotation(hummus@assays$peaks) <- genome_annotations
# Load TF motifs from JASPAR2020 and chromVARmotifs in hummus object
hummus@motifs_db <- get_tf2motifs() # by default human motifs
```


## 2. Construction of the multilayer
### Compute bipartites and add it to hummus object

#### 2.1. TF - peaks bipartite reconstruction
```{r}
hummus <- bipartite_tfs2peaks(
              hummus_object = hummus,
              tf_expr_assay = "RNA", # use to filter TF on only expressed TFs,
                                     # if NULL, all TFs with motifs are used
              peak_assay = "peaks",
              tf_multiplex_name = "TF",
              genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
              )
```

#### 2.2. Genes - peaks bipartite reconstruction
```{r}
hummus <- bipartite_peaks2genes(
                      hummus_object = hummus,
                      gene_assay = "RNA",
                      peak_assay = "peaks",
                      store_network = FALSE,
                      )
```

### Compute layer networks and add it to hummus object
To illustrate the possible level of parameters specifications,
each network will be reconstruct with different personnalisation levels.

#### 2.3. Compute the TF network from OmniPath database
```{r}
hummus <- compute_tf_network(hummus,
                            gene_assay = "RNA", # default = None ;
                                                #If a assay is provided,
                                                # only the TFs that are present
                                                # will be considered
                            verbose = 1,
                            multiplex_name = "TF",
                            tf_network_name = "TF_network")
```


#### 2.4. Compute gene network from scRNA-seq w/ GENIE3
```{r}
hummus <- compute_gene_network(
              hummus,
              gene_assay = "RNA",
              method = "GENIE3",
              verbose = 1,
              number_cores = 5, # GENIE3 method can be ran
                                # parallelised on multiple cores
              store_network = FALSE, # by default : FALSE
                                     # each network can be saved 
                                     # when computed with hummus
              output_file = "gene_network.tsv")
```

#### 2.5. Compute the peak network from scATAC-seq w/ Cicero
```{r}
hummus <- compute_atac_peak_network(hummus,
              atac_assay = "peaks",
              verbose = 1,
              genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
              store_network = FALSE)  
```


## Analyse of the multilayer and definition of GRN
#### 3.1. Save the mulilayer in a classic hierarchical structure
```{r}
save_multilayer(hummus = hummus, "chen_multilayer")
```

#### 3.2. Define GRN
```{r}
grn <- define_grn(
  hummus,
  multilayer_f = "chen_multilayer",
  njobs = 5
  )
```
```{r}
head(grn)
```

#### 3.3. Retrieve enhancers
```{r}
enhancers <- define_enhancers(
  hummus,
  multilayer_f = "chen_multilayer",
  njobs = 5
  )
```
```{r}
head(enhancers)
```

#### 3.4. Retrieve binding regions
```{r}
binding_regions <- define_binding_regions(
  hummus,
  multilayer_f = "chen_multilayer",
  njobs = 5
  )
```
```{r}
head(binding_regions)
```

#### 3.5. Retrieve target genes
```{r}
target_genes <- define_target_genes(
  hummus,
  multilayer_f = "chen_multilayer",
  njobs = 5
  )
```
```{r}
head(target_genes)
```